text/3802-assume-bound.md

```diff
@@ -0,0 +1,92 @@
+- Feature Name: `assume_bound`
+- Start Date: 2025-04-21
+- RFC PR: [rust-lang/rfcs#3802](https://github.com/rust-lang/rfcs/pull/3802)
+- Rust Issue: N/A
+
+# Summary
+[summary]: #summary
+
+This feature allows to assume trait bounds on generics so that the caller don't has to proof them pre-monomorph.
+
+# Motivation
+[motivation]: #motivation
+
+I am currently writing a lot of generic magic again.
+To be more concrete, I am making a framework where lots of functions with generic parameters call eachother.
+There, I pass some `impl Key` values, which hold an assoicated type and proof that the generic, which is passed down, implements `Has<T>`, where T is that associated bound.
+Now I don't really want to specify that requirement on every function, as that is cumbersome and hard to maintain. 
+Thus I propose to let me have my unsafe fun to assume trait bounds are fulfilled in the places I am sure about.
+
+Another use case is for complex higher ranked bounds, where a small helper function could be used to hint the compiler that a trait is really implemented.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When implementing a function with a `where` clause, like this one:
```

Comment:
what about non-functions

```rust
trait Foo<U> where #[unsafe(assume)] U: Iterator {
    type Item where #[unsafe(assume)] <U as Iterator>::Item: Into<u16>;
}

impl Foo<U> for U where #[unsafe(assume)] U: Future {
    type Item = U::Output;
}
```

Reply:
:0 interesting idea, should probably also work on those..

Reply:
That seems like it would be much more difficult to implement/develop coherent semantics for.